libs/PopupRequest.js

```diff
@@ -1,140 +1,91 @@
-                   Object.keys(this.request.query.DATA.CUSTOM_DATA).forEach(function (varName) {
-                       try{
-                           that.CustomDataParmList.push(new CallCustomParam(varName,that.request.query.DATA.CUSTOM_DATA[varName],false));
-                       }catch (e) {
-                           console.error("Failed adding CUSTOM_DATA parameter ",varName);
-                       }
-                   })
-               }
-           }
-       }catch (e) {
-           console.error("parseRequest failed ",e)
-       }
-    }
-    Done(){
-        this.done =true;
-        this.reply.send(this.Result);
-    }
-    async DoPopupLogic() {
-        await this.popupLogic(this);
-        if(!this.done)this.Done()
+    this.clearActionModule();
+    this.requireActionModule();
+  }
 
+  parseRequest() {
+    try {
+      if (this.request.body) {
+        this.parseRequestToObject(this.request.body);
+        this.parseCustomData(this.request.body.CUSTOM_DATA);
+      } else if (this.request.query) {
+        this.parseRequestToObject(this.request.query);
+        this.parseCustomData(this.request.query.CUSTOM_DATA);
+      }
+    } catch (err) {
+      console.error("parseRequest failed ", err);
     }
-    ApprovePopup(approveUrlPath) {
-        this.done = true;
-        const protocol = this.request.protocol || 'http://';
-        const host = this.request.hostname;
-        const popupPath = this.request.raw.originalUrl.split('?')[0];
-        const query = { ...this, request: undefined, reply: undefined, Result: undefined, popupURL: protocol + host + popupPath };
-        this.Result["URL"] = protocol + host + '/PopupApprove/' + approveUrlPath + '?&data=' + jwt.sign(query, jwtKey);
-        this.reply.send(this.Result);
+  }
+
+  ApprovePopup(approveUrlPath) {
+    this.done = true;
+
+    const protocol = this.request.protocol || 'http://';
+    const host = this.request.hostname;
+    const popupPath = this.request.raw.originalUrl.split('?')[0];
+    const query = {
+      ...this,
+      request: undefined,
+      reply: undefined,
+      Result: undefined,
+      popupURL: protocol + host + popupPath
+    };
+
+    try {
+      this.Result["URL"] = protocol + host + '/PopupApprove/' + approveUrlPath + '?&data=' + jwt.sign(query, this.config.jwtKey);
+    } catch (err) {
+      console.error("Faled to encode data to jwt", err);
     }
-    // Call Custom Param Functions  Start
-    SetParam(parmName,paramValue){
-        let params = this.CustomDataParmList.filter(function (p) {return p.Name ===parmName });
-        if(params.length>0){
-            params.forEach(function (param) {
-                param.Update(paramValue)
-            })
-        }else{
-            this.CustomDataParmList.push(new CallCustomParam(parmName,paramValue,true));
-        }
 
+    this.reply.send(this.Result);
+  }
+  // Call Custom Param Functions  Start
+  SetParam(parmName, paramValue) {
```

Comment:
DRY

`CallRequest.js` is using the same code

Move `SetParam()` method to the **Request** main class, and check if there any same repeating methods